### PR TITLE
fix #251 add DataValidation

### LINKED ISSUE #251: Rollback Trx No DataValidation
````
public abstract class AbstractUndoExecutor {
        public void executeOn(Connection conn) throws SQLException {
                dataValidation(conn);
                xxxxxxxx
        }

        protected void dataValidation(Connection conn) throws SQLException {
                // Validate if data is dirty.
        }
}
````

对于数据回滚校验，我想通过 重写 struct 下实体类的 equals 方法实现一版，并扩展 force rollback 选项。

### PR DESCRIPTION
**Pull request by IS908**
* issue #251 add DataValidation

### Ⅰ. Describe what this PR did

For data rollback validation, not implement yet !

### Ⅱ. Does this pull request fix one issue?

fix #251 

### Ⅲ. Describe how to verify it

verify in unitTest:  
com.alibaba.fescar.rm.datasource.undo.UndoExecutorTest

### Ⅳ. Special notes for reviews

check method equals(Object) and hashCode() for more cases.

### CODE REVIEW

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/AbstractUndoExecutor.java [hunk 1]
```diff
@@ -89,7 +87,69 @@ protected void undoPrepare(PreparedStatement undoPST, ArrayList<Field> undoValue
...
+    // Validate if data is dirty.
     protected void dataValidation(Connection conn) throws SQLException {
-        // Validate if data is dirty.
+        String tableName = sqlUndoLog.getTableName();
+        List<Field> pkRows = sqlUndoLog.getAfterImage().pkRows();
+        String querySQL = buildQueryCurrentImageSQL(tableName, pkRows);
+        TableRecords currentImage = getCurrentImage(conn, querySQL, pkRows);
+        currentImage.setTableName(tableName);
+        if (!currentImage.equals(sqlUndoLog.getAfterImage()))
+            throw new RuntimeException("rollback failed! data is dirty!");
```
[the remark this answers is not in the document]
Comment (IS908) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/AbstractUndoExecutor.java]: Thx!

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/AbstractUndoExecutor.java [hunk 2]
```diff
@@ -89,7 +87,69 @@ protected void undoPrepare(PreparedStatement undoPST, ArrayList<Field> undoValue
...
+    }
+
+    protected String buildQueryCurrentImageSQL(String tableName, List<Field> pkRows) {
+        StringBuilder sb = new StringBuilder("SELECT * FROM ");
```
[the remark this answers is not in the document]
Comment (IS908) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/AbstractUndoExecutor.java]: Yes, but when rollback deleteSQL. In this case,  after-image is empty. Should I use star to select all columns.
````
StringBuilder sb = new StringBuilder("SELECT ");
List<Row> rows = sqlUndoLog.getAfterImage().getRows();
if (rows.size() < 1) {
    sb.append(" * ");   // case: afterImage is empty (eg. deleteSQL)
} else {
    Row row = rows.get(0);
    boolean first = true;
    for (Field field : row.getFields()) {
        if (first) {
            first = false;
        } else {
            sb.append(", ");
        }
        sb.append(field.getName());
    }
}
````

[the remark this answers is not in the document]
Comment (sharajava) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/AbstractUndoExecutor.java]: In my opinion, dirty-check for DELETE should be with before-image:

Check if there are records existing with same PK values recorded in before-image.

[the remark this answers is not in the document]
Comment (IS908) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/AbstractUndoExecutor.java]: yes, I did like this: 

For Insert/Update, get selected columns according to after-image.
And For Delete override  as follows:
````
String tableName = sqlUndoLog.getTableName();
TableRecords beforeImage = sqlUndoLog.getBeforeImage();
List<Row> rows = beforeImage.getRows();
if (rows.size() < 1) return;            // if records is empty
List<Field> pkRows = beforeImage.pkRows();
String querySQL = buildQueryCurrentImageSQL(rows.get(0), tableName, pkRows);
TableRecords currentImage = getCurrentImage(conn, querySQL, pkRows);
currentImage.setTableName(tableName);
if (currentImage.getRows().size() > 0) {
	DataSourceManager.get().branchReport(xid, branchId, BranchStatus.PhaseTwo_RollbackFailed_Unretryable, FrameworkErrorCode.RollbackDirty.errMessage);
	throw new FrameworkException(FrameworkErrorCode.RollbackDirty);
}
````

````
Check if there are records existing with same PK values recorded in before-image.
I can not get the point. Is there any examples?
````

**Pull request closed without merging**